### PR TITLE
feat: selector

### LINKED ISSUE #6: Move `find` function (formally know as selector)
As a plugin author, I want to get an SCL Element based on a given identity string.

**Requirements:**
- move find as is
- export `query(root:XMLDocument | Element, tagName: string, identity: string): Element | null`

**Additional Information:**
It is included in `open-scd` already and can be moved, including tests.

### PR DESCRIPTION
**Pull request by JakobVogelsang**
Closes #6 

This time i did touch `IEDName` and `ProtNS`. Both were not tested and for no good reason. Along the way I have increased the test coverage on some test cases such as ExtRef with intAddr and w/o and others. 

The test is running some time as we do the test for all element in the validScl file. I then also promoses to cover a lot of edge cases

### CODE REVIEW

File: utils/selector.spec.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+const doc = new DOMParser().parseFromString(validScl, 'application/xml')!;
+
+describe('Function returning query selector', () => {
+  it('returns negation pseudo-class for identity of type NaN', () =>
+    expect(selector('Association', NaN)).to.equal(':not(*)'));
+
+  it('returns tagName with non SCL tag', () =>
+    expect(selector('LNodeSpec', 'someLNodeSpec')).to.equal('LNodeSpec'));
+
+  it('returns correct selector for all tags except IEDName and ProtNs', () => {
```
Comment (danyill) [inline utils/selector.spec.ts:16]: Are these really an exception now?

**Review by danyill — COMMENTED** [review]
Thank you for moving this important functionality, I have tried to take a look at this functionality in a few places in detail and it looks good.

I have also tried to integrate into `oscd-subscriber-later-binding` and I have quite a range of test failures after I do that which I have not yet found time to look at and if you can wait I would like to examine. I use both identity and selector as part of the UI so I feel that would be worth investigating.

And also I would like to integrate this into openscd/open-scd and check that tests pass and prepare a PR for this so that we can immediately benefit from this library.

**Comment by danyill** [comment]
(At least) part of the reason for my tests failing is that the ExtRef identity function in this library is different to that in open-scd.

In open-scd we have:

```ts
function extRefIdentity(e: Element): string | number {
  if (!e.parentElement) return NaN;
  const parentIdentity = identity(e.parentElement);
  const iedName = e.getAttribute('iedName');
  const intAddr = e.getAttribute('intAddr');
  const intAddrIndex = Array.from(
    e.parentElement.querySelectorAll(`ExtRef[intAddr="${intAddr}"]`)
  ).indexOf(e);
  if (intAddr) return `${parentIdentity}>${intAddr}[${intAddrIndex}]`;
  const [
```

Here we have:

https://github.com/openscd/oscd-scl/blob/2df4a182cdabf47ed59040d0621cd481312c0e1e/utils/identity.ts#L86-L97

The handling of `intAddr` is different as per https://github.com/openscd/open-scd/issues/1179

File: utils/selector.spec.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        if (element)
+          expect(element).to.satisfy(
+            // eslint-disable-next-line no-shadow
+            (element: Element) =>
+              element === doc.querySelector(selector(tag, identity(element)))
+          );
+      });
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (ca-d) [reply utils/selector.spec.ts:32]: Yes, these seem very good to have! They could also catch #14 if the test data was extended to a case with ExtRefs sharing an intAddr.

[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply utils/selector.spec.ts:32]: Thank you. I will add one more `ExtRef` element covering the edge case @danyill triggered here.

File: utils/selector.spec.ts (hunk 1 above)
Comment (JakobVogelsang) [reply utils/selector.spec.ts:16]: True, thank you!

**Comment by JakobVogelsang** [comment]
I decided to close this PR. We cannot resolve some element purly based selector string. Those are in particular `ExtRef` with `intAddr`, `P`, `IEDName`, `ProtNs` and `Val`. All of those must be indexes for the one or the other reason. This conflicts with selectors. The thing that we could use is `_nth-child(n of S)`, but I decided against it as this is a too new feature.

**Comment by JakobVogelsang** [comment]
Please look into this PR for a better way to get SCL element based on an identity string, https://github.com/openscd/open-scd/pull/1285. There is also a branch in preparation that does transport it to this repository. https://github.com/JakobVogelsang/oscd-scl/tree/find
